File: android/app/src/main/java/app/releaf/mobile/ui/components/ReLeafStrip.kt
/*
 * ReLeafStrip.kt
 *
 * Two-tile strip that sits between the CaptureTabBar and the AT A
 * GLANCE grid on the page Overview. Reuses the same card vocabulary
 * as StatGrid so it reads as a section the page always had — not a
 * new component class.
 *
 * Left tile  — SHEETS SAVED + tree silhouette in the bottom-right
 * Right tile — FOREST + five-segment progress dotline
 *
 * Tapping the "RE-LEAF" eyebrow above the tiles fires `onShowDetail`,
 * which the parent wires to the PaperSavedSheet. Eyebrow is the only
 * tap target; the tiles themselves are read-only.
 *
 * Mirrors `ReLeafStrip.swift` (iOS).
 */

package app.releaf.mobile.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Info
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography
import app.releaf.mobile.ui.theme.ReleafImpact

@Composable
fun ReLeafStrip(
    impact: ReleafImpact,
    modifier: Modifier = Modifier,
    onShowDetail: () -> Unit = {},
    /** Optional override for the eyebrow + SHEETS SAVED label
     *  tint. Nil → strip stays in its default themeGreenDeep
     *  tone. Tree silhouette + progress dots intentionally stay
     *  green: those visuals stand for "trees" semantically and
     *  shouldn't morph with notebook color. */
    accentOverride: androidx.compose.ui.graphics.Color? = null,
) {
    val eyebrowTint = accentOverride ?: AppColors.ThemeGreenDeep
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(AppSpacing.s2),
    ) {
        // Eyebrow + info icon. Tappable area is the whole row; the
        // tiles below are read-only.
        Row(
            modifier = Modifier
                .clip(RoundedCornerShape(AppRadius.sm))
                .clickable(onClick = onShowDetail)
                .padding(vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text(
                text  = "RE-LEAF",
                style = AppTypography.Eyebrow,
                color = eyebrowTint,
            )
            Icon(
                imageVector        = Icons.Filled.Info,
                contentDescription = "Show how paper saved is counted",
                tint               = eyebrowTint.copy(alpha = 0.55f),
                modifier           = Modifier.size(11.dp),
            )
        }

        Row(horizontalArrangement = Arrangement.spacedBy(AppSpacing.s2)) {
            SheetsSavedTile(
                value      = impact.formattedSheets,
                accentTint = eyebrowTint,
                modifier   = Modifier.weight(1f),
            )
            ForestTile(
                impact   = impact,
                modifier = Modifier.weight(1f),
            )
        }
    }
}

// MARK: - SHEETS SAVED tile

@Composable
private fun SheetsSavedTile(
    value: String,
    accentTint: androidx.compose.ui.graphics.Color,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(AppRadius.lg))
            .background(AppColors.CardSolid)
            .border(1.dp, AppColors.BorderDefault, RoundedCornerShape(AppRadius.lg)),
    ) {
        Column(
            modifier = Modifier.padding(AppSpacing.s4),
            verticalArrangement = Arrangement.spacedBy(AppSpacing.s2),
        ) {
            Text(
                text  = "SHEETS SAVED",
                style = AppTypography.Eyebrow,
                color = accentTint,
            )
            Text(
                text     = value,
                style    = AppTypography.StatNumber.copy(fontFamily = androidx.compose.ui.text.font.FontFamily.Serif),
                color    = AppColors.TextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Clip,
            )
            Text(
                text  = "paper not printed",
                style = AppTypography.Tag,
                color = AppColors.TextTertiary,
            )
        }

        TreeSilhouette(
            modifier = Modifier
                .size(width = 56.dp, height = 52.dp)
                .offset(x = 6.dp, y = 6.dp)
                .align(Alignment.BottomEnd)
        )
    }
}

// MARK: - FOREST tile

@Composable
private fun ForestTile(
    impact: ReleafImpact,
    modifier: Modifier = Modifier,
) {
    val readout = impact.treeReadout
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(AppRadius.lg))
            .background(AppColors.CardSolid)
            .border(1.dp, AppColors.BorderDefault, RoundedCornerShape(AppRadius.lg)),
    ) {
        Column(
            modifier = Modifier.padding(AppSpacing.s4),
            verticalArrangement = Arrangement.spacedBy(AppSpacing.s2),
        ) {
            Text(
                text  = "FOREST",
                style = AppTypography.Eyebrow,
                color = AppColors.TextSecondary,
            )
            Text(
                text     = readout.number,
                style    = AppTypography.StatNumber.copy(fontFamily = androidx.compose.ui.text.font.FontFamily.Serif),
                color    = AppColors.TextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Clip,
            )
            Text(
                text  = readout.unit,
                style = AppTypography.Tag,
                color = AppColors.TextTertiary,
            )
            ForestProgressDots(treeFraction = impact.treeFraction)
        }
    }
}

// MARK: - Five-segment progress dotline

private const val SegmentCount = 5

@Composable
private fun ForestProgressDots(treeFraction: Double) {
    Row(
        modifier              = Modifier.fillMaxWidth().padding(top = 2.dp),
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        for (idx in 0 until SegmentCount) {
            val threshold = (idx + 1).toDouble() / SegmentCount.toDouble()
            val color = when {
                treeFraction >= threshold && threshold == 1.0 -> AppColors.Green
                treeFraction >= threshold && threshold >= 0.8 -> AppColors.ThemeGreenDeep
                treeFraction >= threshold                     -> AppColors.ThemeGreenPrimary
                else                                          -> AppColors.ThemeGreenPrimary.copy(alpha = 0.25f)
            }
            Box(
                modifier = Modifier
                    .weight(1f)
                    .height(5.dp)
                    .clip(CircleShape)
                    .background(color)
            )
        }
    }
}

// MARK: - Tree silhouette

@Composable
private fun TreeSilhouette(modifier: Modifier = Modifier) {
    val trunk      = Color(0xFF8B7355)
    val midGreen   = AppColors.ThemeGreenPrimary
    val deepGreen  = AppColors.ThemeGreenDeep
    Canvas(modifier = modifier) {
        val w = size.width
        val h = size.height
        // Trunk
        drawRoundRect(
            color    = trunk,
            topLeft  = Offset(w * 0.46f, h * 0.62f),
            size     = Size(w * 0.08f, h * 0.34f),
            cornerRadius = androidx.compose.ui.geometry.CornerRadius(1.5f, 1.5f),
        )
        // Canopy back
        drawOval(
            color   = midGreen,
            topLeft = Offset(w * 0.10f, h * 0.05f),
            size    = Size(w * 0.80f, h * 0.65f),
        )
        // Canopy right
        drawOval(
            color   = deepGreen,
            topLeft = Offset(w * 0.45f, 0f),
            size    = Size(w * 0.55f, h * 0.50f),
        )
        // Canopy left
        drawOval(
            color   = deepGreen,
            topLeft = Offset(0f, h * 0.05f),
            size    = Size(w * 0.50f, h * 0.45f),
        )
    }
}

// MARK: - Preview

@Preview(showBackground = true, backgroundColor = 0xFFF5EEDF, widthDp = 360)
@Composable
private fun ReLeafStripPreview() {
    Column(
        modifier              = Modifier.padding(AppSpacing.s4).background(AppColors.Canvas),
        verticalArrangement   = Arrangement.spacedBy(AppSpacing.s4),
    ) {
        ReLeafStrip(
            impact = ReleafImpact.from(
                photos = 2, voiceNotes = 1, todoItems = 1,
                scans = 2, contacts = 1, places = 1, notes = 1,
            )
        )
        ReLeafStrip(
            impact = ReleafImpact.from(
                photos = 84, voiceNotes = 24, todoItems = 47,
                scans = 52, contacts = 8, places = 14, notes = 38,
            )
        )
    }
}

File: apps/releaf/ios/Releaf/DesignSystem/Components/ReLeafStrip.swift
/*
 * ReLeafStrip.swift
 *
 * Two-tile strip that sits between the CaptureTabBar and the AT A
 * GLANCE grid on the page Overview. Reuses the same card vocabulary
 * as StatTile so it reads as a section the page always had — not a
 * new component class.
 *
 * Left tile  — SHEETS SAVED + tree silhouette in the bottom-right
 * Right tile — FOREST + five-segment progress dotline
 *
 * Tapping the "RE-LEAF" eyebrow above the tiles fires `onShowDetail`,
 * which the parent wires to the PaperSavedSheet. Eyebrow is the only
 * tap target; the tiles themselves are read-only.
 */

import SwiftUI

public struct ReLeafStrip: View {
    public let impact: ReleafImpact
    public let onShowDetail: () -> Void
    /// Optional override for the eyebrow + SHEETS SAVED label
    /// tint. When nil, the strip stays in its default
    /// themeGreenDeep tone — appropriate for a Re-Leaf metric.
    /// PageDetail passes the parent-notebook color when set so
    /// the strip reads as part of the surface family.
    /// The tree silhouette + progress dots intentionally stay
    /// green: those visuals stand for "trees" semantically and
    /// shouldn't morph with notebook color.
    public let accentOverride: Color?

    public init(
        impact: ReleafImpact,
        onShowDetail: @escaping () -> Void = {},
        accentOverride: Color? = nil
    ) {
        self.impact = impact
        self.onShowDetail = onShowDetail
        self.accentOverride = accentOverride
    }

    public var body: some View {
        let eyebrowTint = accentOverride ?? AppColors.themeGreenDeep
        VStack(alignment: .leading, spacing: AppSpacing.s2) {
            Button(action: onShowDetail) {
                HStack(spacing: AppSpacing.s1) {
                    Text("RE-LEAF")
                        .font(AppText.eyebrow)
                        .tracking(AppLetterSpacing.eyebrow)
                        .foregroundStyle(eyebrowTint)
                    Image(systemName: "info.circle")
                        .font(.system(size: 10))
                        .foregroundStyle(eyebrowTint.opacity(0.55))
                }
            }
            .buttonStyle(.plain)
            .accessibilityHint("Show how paper saved is counted")

            HStack(spacing: AppSpacing.s2) {
                SheetsSavedTile(value: impact.formattedSheets, accentTint: eyebrowTint)
                ForestTile(impact: impact)
            }
        }
    }
}

// MARK: - SHEETS SAVED tile

private struct SheetsSavedTile: View {
    let value: String
    /// Eyebrow tint. Defaults to the green theme; ReLeafStrip
    /// passes through the override when one was supplied so the
    /// tile's eyebrow matches the parent strip's eyebrow.
    let accentTint: Color

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            VStack(alignment: .leading, spacing: AppSpacing.s2) {
                Text("SHEETS SAVED")
                    .font(AppText.eyebrow)
                    .tracking(AppLetterSpacing.eyebrow)
                    .foregroundStyle(accentTint)
                Text(value)
                    .font(.system(size: 32, design: .serif))
                    .foregroundStyle(AppColors.textPrimary)
                    .lineLimit(1)
                    .minimumScaleFactor(0.6)
                Text("paper not printed")
                    .font(AppText.tag)
                    .foregroundStyle(AppColors.textTertiary)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(AppSpacing.s4)

            TreeSilhouette()
                .frame(width: 56, height: 52)
                .padding(.trailing, -6)
                .padding(.bottom, -6)
                .opacity(0.95)
                .accessibilityHidden(true)
        }
        .frame(maxWidth: .infinity)
        .background(AppColors.cardSolid)
        .overlay(
            RoundedRectangle(cornerRadius: AppRadius.lg, style: .continuous)
                .stroke(AppColors.borderDefault, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: AppRadius.lg, style: .continuous))
        .appShadow(.sm)
    }
}

// MARK: - FOREST tile

private struct ForestTile: View {
    let impact: ReleafImpact

    var body: some View {
        let readout = impact.treeReadout
        VStack(alignment: .leading, spacing: AppSpacing.s2) {
            Text("FOREST")
                .font(AppText.eyebrow)
                .tracking(AppLetterSpacing.eyebrow)
                .foregroundStyle(AppColors.textSecondary)
            Text(readout.number)
                .font(.system(size: 32, design: .serif))
                .foregroundStyle(AppColors.textPrimary)
                .lineLimit(1)
                .minimumScaleFactor(0.5)
            Text(readout.unit)
                .font(AppText.tag)
                .foregroundStyle(AppColors.textTertiary)
            ForestProgressDots(treeFraction: impact.treeFraction)
                .padding(.top, AppSpacing.s1)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(AppSpacing.s4)
        .background(AppColors.cardSolid)
        .overlay(
            RoundedRectangle(cornerRadius: AppRadius.lg, style: .continuous)
                .stroke(AppColors.borderDefault, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: AppRadius.lg, style: .continuous))
        .appShadow(.sm)
    }
}

// MARK: - Five-segment progress dotline

private struct ForestProgressDots: View {
    /// 0.0 → empty, 1.0 → all five segments lit.
    let treeFraction: Double

    private static let segmentCount = 5

    var body: some View {
        HStack(spacing: 3) {
            ForEach(0..<Self.segmentCount, id: \.self) { idx in
                let threshold = Double(idx + 1) / Double(Self.segmentCount)
                Capsule()
                    .fill(color(for: threshold))
                    .frame(height: 5)
            }
        }
    }

    private func color(for threshold: Double) -> Color {
        if treeFraction >= threshold {
            // Filled segments shade darker as you move right — the
            // visual hint is "growing toward a full tree".
            switch threshold {
            case 1.0:  return AppColors.green
            case 0.8:  return AppColors.themeGreenDeep
            default:   return AppColors.themeGreenPrimary
            }
        }
        return AppColors.themeGreenPrimary.opacity(0.25)
    }
}

// MARK: - Tree silhouette

private struct TreeSilhouette: View {
    var body: some View {
        Canvas { context, size in
            // Trunk
            let trunkRect = CGRect(
                x: size.width * 0.46,
                y: size.height * 0.62,
                width: size.width * 0.08,
                height: size.height * 0.34
            )
            context.fill(
                Path(roundedRect: trunkRect, cornerRadius: 1.5),
                with: .color(Color(hex: 0x8B7355))
            )

            // Canopy — three overlapping ellipses, deeper greens behind.
            let canopyBack = Path(ellipseIn: CGRect(
                x: size.width * 0.10,
                y: size.height * 0.05,
                width: size.width * 0.80,
                height: size.height * 0.65
            ))
            context.fill(canopyBack, with: .color(AppColors.themeGreenPrimary))

            let canopyRight = Path(ellipseIn: CGRect(
                x: size.width * 0.45,
                y: size.height * 0.0,
                width: size.width * 0.55,
                height: size.height * 0.50
            ))
            context.fill(canopyRight, with: .color(AppColors.themeGreenDeep))

            let canopyLeft = Path(ellipseIn: CGRect(
                x: size.width * 0.0,
                y: size.height * 0.05,
                width: size.width * 0.50,
                height: size.height * 0.45
            ))
            context.fill(canopyLeft, with: .color(AppColors.themeGreenDeep))
        }
    }
}

// MARK: - Preview

#if DEBUG
struct ReLeafStrip_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: AppSpacing.s4) {
            ReLeafStrip(
                impact: ReleafImpact(
                    photos: 2, voiceNotes: 1, todoItems: 1,
                    scans: 2, contacts: 1, places: 1, notes: 1
                )
            )
            ReLeafStrip(
                impact: ReleafImpact(
                    photos: 84, voiceNotes: 24, todoItems: 47,
                    scans: 52, contacts: 8, places: 14, notes: 38
                )
            )
        }
        .padding(AppSpacing.s4)
        .background(AppColors.canvas)
        .previewDisplayName("ReLeafStrip — single page vs aggregate")
    }
}
#endif
